src/hbb/processors/categorizer.py

```diff
@@ -277,37 +356,67 @@ def process_shift(self, events, shift_name):
         selection.add("isvbf", isvbf)
         selection.add("notvbf", isnotvbf)
 
-        goodmuon = good_muons(events.Muon)
+        muons = correct_muons(events.Muon, events, self._year, isRealData)
+        if not shift_name == "nominal" and "Muon" in shift_name:
+            var, direction = shift_name.split("_")
+            self._mupt_type = f"{mupt_variations[var]}_{direction.lower()}"
+
+        goodmuon = good_muons(muons, self._mupt_type)
         nmuons = ak.num(goodmuon, axis=1)
         leadingmuon = ak.firsts(goodmuon)
+        ttbarmuon = ak.firsts(goodmuon[getattr(goodmuon, self._mupt_type) > 55.])
+            #low pt muons break sf (lower bound 15GeV)
 
         goodelectron = good_electrons(events.Electron)
         nelectrons = ak.num(goodelectron, axis=1)
 
         selection.add("noleptons", (nmuons == 0) & (nelectrons == 0))
         selection.add("onemuon", (nmuons == 1) & (nelectrons == 0))
-        selection.add("muonkin", (leadingmuon.pt > 55.0) & (abs(leadingmuon.eta) < 2.1))
+        selection.add("muonkin", (getattr(leadingmuon, self._mupt_type) > 55.0) & (abs(leadingmuon.eta) < 2.1))
         selection.add("muonDphiAK8", abs(leadingmuon.delta_phi(candidatejet)) > 2 * np.pi / 3)
 
         goodphotons = good_photons(events.Photon)
         nphotons = ak.num(goodphotons, axis=1)
         leadingphoton = ak.firsts(goodphotons)
         ntightphotons = ak.num(tight_photons(events.Photon), axis=1)
+        vgammaphoton = ak.firsts(tight_photons(events.Photon))
+            #low pt photons break sf (lower bound 20GeV)
+            #TODO confirm selection of this block with Gabi - leading photon still from goodphotons collection?
 
         selection.add("onephoton", (nphotons == 1))
         selection.add("atleastonephoton", (ntightphotons >= 1))
         selection.add("passphotonveto", (nphotons == 0))
 
         gen_variables = {}
+        btag_SF = 1.
         if isRealData:
             genflavor = ak.zeros_like(candidatejet.pt)
             genBosonPt = ak.zeros_like(candidatejet.pt)
         else:
-            weights_dict, totals_temp = self.add_weights(
+            #signal regions
```

Comment:
I suggest that instead of using `signal, mu, egamma` to get the weights you instead use a single function to add the weights except the btagSFs, and then another function to add the btagSFs for  `ak4_opphem_ak8, ak4_outside_ak8`, and finally two other functions to add the weights for photons and for muons. What do you think?